File: go-solutions/1657_close_strings.go
package leetcode

import (
	"reflect"
	"slices"
)

func CloseStrings(word1, word2 string) bool {
	word1Len := len(word1)
	word2Len := len(word2)
	if word1Len != word2Len {
		return false
	}
	word1Map, word2Map := make([]uint8, 26), make([]uint8, 26)
	for i := 0; i < word1Len; i++ {
		word1Map[word1[i]-'a']++
		word2Map[word2[i]-'a']++
	}
	for i := 0; i < 26; i++ {
		if word1Map[i] > 0 && word2Map[i] == 0 {
			return false
		}
		if word2Map[i] > 0 && word1Map[i] == 0 {
			return false
		}
	}
	slices.Sort(word1Map)
	slices.Sort(word2Map)
	return reflect.DeepEqual(word1Map, word2Map)
}

File: ts-solutions/src/1657-close-strings.ts
export const closeStrings = (word1: string, word2: string) => {
	if (word1.length != word2.length) {
		return false
	}
	const word1Map = new Array<number>(26).fill(0)
	const word2Map = new Array<number>(26).fill(0)
	let word1Bit = 0
	let word2Bit = 0
	for (let i = 0; i < word1.length; i++) {
		const char1 = word1.charCodeAt(i) - 97
		const char2 = word2.charCodeAt(i) - 97
		word1Map[char1]++
		word2Map[char2]++
		word1Bit = word1Bit | (1 << char1)
		word2Bit = word2Bit | (1 << char2)
	}
	if (word1Bit !== word2Bit) {
		return false
	}
	word1Map.sort()
	word2Map.sort()
	for (let i = 0; i < 26; i++) {
		if (word1Map[i] !== word2Map[i]) {
			return false
		}
	}
	return true
}
